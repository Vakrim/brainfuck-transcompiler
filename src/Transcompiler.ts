import { Address } from "./Address";
import { Memory } from "./Memory";
import { Scope } from "./Scope";
import { TemporaryVariable } from "./TemporaryVariable";
import { Variable } from "./Variable";

export class Transcompiler {
  #cursorPosition: Address = 0 as Address;
  #memory: Memory;
  #scope: Scope;
  #code: string;
  #commentBuffer: string;

  constructor() {
    this.#cursorPosition = 0 as Address;
    this.#memory = new Memory();
    this.#scope = new Scope(null, this.#memory);
    this.#code = "";
    this.#commentBuffer = "";
  }

  declareVariable(name: string) {
    this.#comment(`declare ${name}`);

    this.#scope.declareVariable(name, this.#cursorPosition);
  }

  assignValue(name: string, value: number) {
    this.#comment(`assign ${value} to ${name}`);

    this.#moveToVariable(name);
    this.#setValue(value);
  }

  writeInput(name: string) {
    this.#comment(`write input ${name}`);

    this.#moveToVariable(name);
    this.#outputBrainfuck(",");
  }

  readVariable(name: string) {
    this.#comment(`read ${name}`);

    this.#moveToVariable(name);
    this.#readValue();
  }

  add(to: string, from: string) {
    this.#comment(`add ${from} to ${to}`);

    this.#moveToVariable(from);
    const newFrom = this.#scope.declareTemporaryVariable(this.#cursorPosition);

    this.#loop(() => {
      this.#moveToVariable(newFrom);
      this.#inc();

      this.#moveToVariable(to);
      this.#inc();
    });

    this.#scope.promoteVariable(from, newFrom);
  }

  get code() {
    return this.#code;
  }

  #comment(comment: string) {
    this.#commentBuffer = this.#commentBuffer
      ? [this.#commentBuffer, comment].join("; ")
      : comment;
  }

  #moveToVariable(nameOrVariable: string | Variable | TemporaryVariable) {
    const variable =
      typeof nameOrVariable === "string"
        ? this.#scope.getVariable(nameOrVariable)
        : nameOrVariable;

    this.#moveTo(variable.address);
  }

  #moveTo(address: Address) {
    const difference = address - this.#cursorPosition;
    if (difference < 0) {
      this.#outputBrainfuck("<".repeat(-difference));
    } else if (difference > 0) {
      this.#outputBrainfuck(">".repeat(difference));
    }
    this.#cursorPosition = address;
  }

  #loop(fn: () => void) {
    this.#outputBrainfuck(`[`);
    this.#dec();
    const loopConditionPostion = this.#cursorPosition;
    fn();
    this.#moveTo(loopConditionPostion);
    this.#outputBrainfuck(`]`);
  }

  #outputBrainfuck(bf: string) {
    if (this.#commentBuffer) {
      this.#code += `\n${this.#commentBuffer}\n`;
      this.#commentBuffer = "";
    }
    this.#code += bf;
  }

  #dec() {
    this.#outputBrainfuck(`-`);
  }

  #inc() {
    this.#outputBrainfuck(`+`);
  }

  #setValue(value: number) {
    this.#outputBrainfuck("+".repeat(value));
  }

  #readValue() {
    this.#outputBrainfuck(".");
  }
}
