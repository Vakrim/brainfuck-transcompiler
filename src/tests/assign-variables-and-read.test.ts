import { Transcompiler } from "../Transcompiler";
import { executeCode } from "./execute-code";

describe("assign variables and read", () => {
  it("reads one variable", () => {
    const compiler = new Transcompiler();

    compiler.declareVariable("a");
    compiler.assignValue("a", 97);
    compiler.readVariable("a");

    expect(compiler.code).toMatchSnapshot();

    expect(executeCode(compiler.code).output).toEqual('a');
  });

  it("reads two variable", () => {
    const compiler = new Transcompiler();

    compiler.declareVariable("a");
    compiler.declareVariable("b")

    compiler.assignValue("a", 97);
    compiler.assignValue("b", 98);
    
    compiler.readVariable("b");
    compiler.readVariable("a");

    expect(compiler.code).toMatchSnapshot();

    expect(executeCode(compiler.code).output).toEqual('ba');
  });
});
