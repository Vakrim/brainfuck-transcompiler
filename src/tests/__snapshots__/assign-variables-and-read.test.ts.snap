// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assign variables and read reads one variable 1`] = `
"
declare a; assign 97 to a
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
read a
."
`;

exports[`assign variables and read reads two variable 1`] = `
"
declare a; declare b; assign 97 to a
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
assign 98 to b
>++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
read b
.
read a
<."
`;
